packages/babel-plugin-transform-object-rest-spread/src/index.ts

```diff
@@ -526,37 +536,63 @@ export default declare((api, opts: Options) => {
         if (leftPath.isObjectPattern() && hasRestElement(leftPath)) {
           const nodes = [];
 
-          const refName = path.scope.generateUidBasedOnNode(
-            path.node.right,
-            "ref",
-          );
+          const { right } = path.node;
 
-          nodes.push(
-            t.variableDeclaration("var", [
-              t.variableDeclarator(t.identifier(refName), path.node.right),
-            ]),
-          );
+          let refName;
+
+          if (
+            t.isIdentifier(right) &&
+            (pureGetters ||
+              path.scope.getBinding(right.name)?.constantViolations.every(
+                // declared in loop
```

Comment:
```suggestion
                // If a variable is only reassigned through other variable declarations,
                // we are sure that it is not being reassigned while destructuring it.
```

Other examples:
```js
var a = x;
({ u, v, ...w } = a;
var a = y;
```
(please, add it as a test!)